horovod/mxnet/mpi_ops.py

```diff
@@ -206,9 +216,47 @@ def broadcast_(tensor, root_rank, name=None):
     c_in = tensor.handle
     c_out = tensor.handle
     if isinstance(name, string_types):
-        check_call(MPI_MXNET_LIB_CTYPES.horovod_mxnet_broadcast_async(c_in,
-                   c_out, ctypes.c_int(root_rank), c_str(name)))
+        handle = MPI_MXNET_LIB_CTYPES.horovod_mxnet_broadcast_async(
+            c_in, c_out, ctypes.c_int(root_rank), c_str(name))
     else:
-        check_call(MPI_MXNET_LIB_CTYPES.horovod_mxnet_broadcast_async(c_in,
-                   c_out, ctypes.c_int(root_rank), name))
-    return tensor
+        handle = MPI_MXNET_LIB_CTYPES.horovod_mxnet_broadcast_async(
+            c_in, c_out, ctypes.c_int(root_rank), name)
+
+    _handle_map[handle] = tensor
+    return synchronize(handle)
+
+
+def poll(handle):
+    """
+    Polls an allreduce, allgather or broadcast handle to determine whether
+    underlying asynchronous operation has completed. After `poll()` returns
+    `True`, `synchronize()` will return without blocking.
+
+    Arguments:
+        handle: A handle returned by an allreduce, allgather or broadcast
+        asynchronous operation.
+
+    Returns:
+        A flag indicating whether the operation has completed.
+    """
+    return MPI_MXNET_LIB_CTYPES.horovod_mxnet_poll(handle) != 0
+
+
+def synchronize(handle):
+    """
+    Synchronizes an asynchronous allreduce, allgather or broadcast operation
+    until it's completed. Returns the result of the operation.
+
+    Arguments:
+        handle: A handle returned by an allreduce, allgather or broadcast
+        asynchronous operation.
+
+    Returns:
+        An output tensor of the operation.
+    """
+    if handle not in _handle_map:
+        return
+
+    check_call(MPI_MXNET_LIB_CTYPES.horovod_mxnet_wait_and_clear(handle))
+    output = _handle_map.pop(handle)
```

Comment:
shall we also remove the handle from the _handle_map since it has been cleared in the C++ land?

Reply:
isn't `pop()` already doing this?

Reply:
Yes, you are right.